File: backend/src/main/scala/services/battle/microservices/abilities/objects/skill/SkillOutcomeReason.scala
package services.battle.microservices.abilities.objects.skill

enum SkillOutcomeReason {
  case SkillNotOwned
  case Cooldown
  case MissingTarget
  case OutOfRange
  case InvalidTarget
  case NoDirection
  case Blocked
  case InsufficientStamina
}

object SkillOutcomeReason {
  def wireValue(value: SkillOutcomeReason): String =
    value match {
      case SkillOutcomeReason.SkillNotOwned       => "skill_not_owned"
      case SkillOutcomeReason.Cooldown            => "cooldown"
      case SkillOutcomeReason.MissingTarget       => "missing_target"
      case SkillOutcomeReason.OutOfRange          => "out_of_range"
      case SkillOutcomeReason.InvalidTarget       => "invalid_target"
      case SkillOutcomeReason.NoDirection         => "no_direction"
      case SkillOutcomeReason.Blocked             => "blocked"
      case SkillOutcomeReason.InsufficientStamina => "insufficient_stamina"
    }

  def fromWire(value: String): Option[SkillOutcomeReason] =
    value match {
      case "skill_not_owned"       => Some(SkillOutcomeReason.SkillNotOwned)
      case "cooldown"              => Some(SkillOutcomeReason.Cooldown)
      case "missing_target"        => Some(SkillOutcomeReason.MissingTarget)
      case "out_of_range"          => Some(SkillOutcomeReason.OutOfRange)
      case "invalid_target"        => Some(SkillOutcomeReason.InvalidTarget)
      case "no_direction"          => Some(SkillOutcomeReason.NoDirection)
      case "blocked"               => Some(SkillOutcomeReason.Blocked)
      case "insufficient_stamina"  => Some(SkillOutcomeReason.InsufficientStamina)
      case _                       => None
    }
}

File: frontend/src/objects/battle/microservices/abilities/objects/skill/SkillOutcomeReason.ts
export type SkillOutcomeReason =
  | "skill_not_owned"
  | "cooldown"
  | "missing_target"
  | "out_of_range"
  | "invalid_target"
  | "no_direction"
  | "blocked"
  | "insufficient_stamina";

export function isSkillOutcomeReason(value: unknown): value is SkillOutcomeReason {
  return value === "skill_not_owned" ||
    value === "cooldown" ||
    value === "missing_target" ||
    value === "out_of_range" ||
    value === "invalid_target" ||
    value === "no_direction" ||
    value === "blocked" ||
    value === "insufficient_stamina";
}
